sled-agent/src/services.rs

```diff
@@ -2688,56 +2685,103 @@ impl ServiceManager {
+    // which gives the storage manager an opportunity to validate the zone's
+    // storage configuration against the reality of the current sled.
+    async fn start_omicron_zone(
         &self,
-        existing_zones: &mut BTreeMap<String, RunningZone>,
-        requests: &Vec<OmicronZoneConfigLocal>,
-    ) -> Result<(), Error> {
-        if let Some(name) = requests
-            .iter()
-            .map(|request| request.zone.zone_name())
-            .duplicates()
-            .next()
+        zone: &OmicronZoneConfig,
+        time_is_synchronized: bool,
+        all_u2_pools: &Vec<ZpoolName>,
+    ) -> Result<OmicronZone, Error> {
+        // Ensure the zone has been fully removed before we try to boot it.
+        //
+        // This ensures that old "partially booted/stopped" zones do not
+        // interfere with our installation.
+        self.ensure_removed(&zone).await?;
+
+        // If this zone requires timesync and we aren't ready, fail it early.
+        if zone_requires_timesync(&zone.zone_type) && !time_is_synchronized {
+            return Err(Error::TimeNotSynchronized);
+        }
+
+        // Ensure that this zone's storage is ready.
+        let root = self
+            .validate_storage_and_pick_mountpoint(&zone, &all_u2_pools)
+            .await?;
+
+        let config = OmicronZoneConfigLocal { zone: zone.clone(), root };
+
+        let runtime = self
+            .initialize_zone(
+                ZoneArgs::Omicron(&config),
+                // filesystems=
+                &[],
+                // data_links=
+                &[],
+            )
+            .await?;
+
+        Ok(OmicronZone { runtime, config })
+    }
+
+    // Concurrently attempts to start all zones identified by requests.
+    //
+    // This method is NOT idempotent.
+    //
+    // If we try to start ANY zones concurrently, the result is contained
+    // in the `StartZonesResult` value. This will contain the set of zones which
+    // were initialized successfully, as well as the set of zones which failed
+    // to start.
+    async fn start_omicron_zones(
+        &self,
+        requests: impl Iterator<Item = &OmicronZoneConfig> + Clone,
+        time_is_synchronized: bool,
+        all_u2_pools: &Vec<ZpoolName>,
+    ) -> Result<StartZonesResult, Error> {
+        if let Some(name) =
+            requests.clone().map(|zone| zone.zone_name()).duplicates().next()
         {
             return Err(Error::BadServiceRequest {
                 service: name,
                 message: "Should not initialize zone twice".to_string(),
             });
         }
 
-        let futures = requests.iter().map(|request| {
-            async move {
-                self.initialize_zone(
-                    ZoneArgs::Omicron(request),
-                    // filesystems=
-                    &[],
-                    // data_links=
-                    &[],
-                )
+        let futures = requests.clone().map(|zone| async move {
```

Comment:
I think this clone isn't needed

```suggestion
        let futures = requests.map(|zone| async move {
```

Reply:
Removed!